### DISCUSSION TITLE
How to use nested grids with an object represented by Csla.Blazor.ViewModel

### DISCUSSION THREAD

**Discussion by ecaldentey** [MarimerLLC/csla#4297] [Questions]
Hello,

I have a root BusinessClass that has children represented by BusinessListBase classes whose items are BusinessBase child objects, which in turn have children represented in the same manner and can even have further descendants. In other words, I have relationships where the root object (Grandparent) has children (Parents), their children (Grandchildren), and their children (Great-grandchildren).

In my Blazor application, I define a class Csla.Blazor.ViewModel<GrandparentEdit> viewModel that I load from the database with all its childs in the OnAfterRenderAsync event of the Blazor page.

To display the data of the children, grandchildren, and great-grandchildren in the UI, in the Blazor page, I want to use nested grids or components by level, but I don't know how to assign a list that is a property of the Csla.Blazor.ViewModel because if I directly assign a BusinessListBase, it generates NullReferenceException errors in the CSLA properties of the class, as shown in the example:

![image](https://github.com/user-attachments/assets/3d31e5d4-d519-4e91-ab8f-4db7e5addeef)

If I use the ReadOnly versions of the objects where the Grandparent is a ReadOnlyBase root that has a property ReadOnlyListBase for the list of Parent ReadOnlyBase children, everything works, and when moving between grids, the data is updated.

The problem is with the classes that allow editing, and the solution I have been using so far involves versions of the DTO objects of the classes where I add properties to manage editing such as IsDirty, IsNew, IsDeleted, etc., but obviously, that is not the idea; it is complex to synchronize the DTO with the editable objects to make updates to the database, and the goal is to be able to use all the properties of CSLA.

How can I pass a BusinessListBase to the grid or as a parameter to a Blazor component?
How do I load a Csla.Blazor.ViewModel with the BusinessListBase object?

Thank you!

**Comment by StefanOssendorf** [comment]
Hi. How are you creating the object graph? The screenshot provided suggests you are getting a NRE because the application context is probably not set correctly. This has nothing to-do with the view model as far as I understand the screenshot.

**Reply by ecaldentey** [reply]
The class is created correctly, and the children are loaded properly. The screenshot is taken when the Blazor page is rendering, and the error occurs when trying to assign the lists to the grids. However, the entire graph was loaded correctly from the database, and I can traverse it completely using a series of nested foreach loops.

**Reply by hurcane** [reply]
I am currently in the process of upgrading a Csla 5 project and have seen this error a lot during my steps. I believe that @StefanOssendorf has the right idea for your problem. 

The objects may be instantiated, but they may not have been instantiated in a way that has set the internal Csla.ApplicationContext field. If you are using an enapsulated data portal, then you must instantiate the object through a DataPortal call. If you are using an object factory, the object must be instantiated using one of the CreateInstanceDI methods from an existing instance of Csla.ApplicationContext,

SetProperty accesses the ApplicationContext instance, so you will get a null reference exception if it hasn't been set.

In Visual Studio, try to view the properties of "this" in the Locals window. If you ApplicationContext is not set, you are going to see errors instead of values for every property that is using GetProperty. You can also drill into the non-public properties and see that ApplicationContext is null.

**Comment by rockfordlhotka** [comment]
Is it possible that the data grid control is adding a new item to the grid without going through the data portal? In such a case, the object wouldn't be initialized correctly and can't work.

This is a common issue with data grid controls, as they don't follow any standard approach for creating new objects to add to the grid (and therefore the collection).

**Reply by ecaldentey** [reply]
Hello Rocky,

When I directly assign the BusinessListBase to the grid, there are no problems at first; the grid loads well with the data from the list and displays it perfectly. 

![image](https://github.com/user-attachments/assets/5aa3c4bd-6d19-40d8-980f-2111e00e9aaa)

However, when trying to edit a record in the grid, the error shown in the image occurs.

![image](https://github.com/user-attachments/assets/3a282c7e-62af-4899-a8fa-e83c1f1119ef)


Copilot gives me this suggestion, and I’m not sure if it gives you an insight into where the problem might be.

This error happens because the application is attempting to set a property (ProviderId) on an object that has not been properly initialized, leading to a System.NullReferenceException.
This might be happening because:
1.	The object instance of BusinessDealProductServiceEdit is not fully initialized before accessing the ProviderId property.
2.	The ProviderIdProperty might not be registered correctly, causing the SetProperty method to fail.
3.	There could be an issue with the authorization cache in the VerifyAuthorizationCache method, which is causing the CanWriteProperty method to fail.
To fix this, ensure that:
1.	The BusinessDealProductServiceEdit object is fully instantiated before setting the ProviderId property.
2.	Verify that ProviderIdProperty is correctly registered using RegisterProperty.
3.	Check the implementation of VerifyAuthorizationCache and 

That's why what I do is convert the BusinessListBase into a list of DTOs, and everything works well there, although I lose all CRUD functionality of CSLA and have to manually synchronize with the BusinessListBase before saving to the database using CSLA.

**Reply by StefanOssendorf** [reply]
Can you provide the stack trace from that exception? Maybe there's something to identify the issue.

**Reply by ecaldentey** [reply]
Hello Stefan

This is the StackTrace of the error:

System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=Csla
  StackTrace:
   at Csla.Core.BusinessBase.VerifyAuthorizationCache()
   at Csla.Core.BusinessBase.CanWriteProperty(IPropertyInfo property)
   at Csla.Core.BusinessBase.CanWriteProperty(IPropertyInfo property, Boolean throwOnFalse)
   at Csla.Core.BusinessBase.SetProperty[P](PropertyInfo`1 propertyInfo, P newValue, NoAccessBehavior noAccess)
   at Csla.Core.BusinessBase.SetProperty[P](PropertyInfo`1 propertyInfo, P newValue)
   at ICL.Catodos.ERP.Library.ERP.BusinessDealProductServiceEdit.set_ProviderId(Decimal value) in D:\Desarrollo\ICL\ICL.Catodos.ERP\Solution\ICL.Catodos.ERP\ICL.Catodos.ERP.Library\ERP\BusinessDealProductServiceEdit.Generated.cs:line 51

![image](https://github.com/user-attachments/assets/7ed91144-f5f6-4efd-84b7-7879510f32ae)

**Reply by knox007** [reply]
Have you solve this problem?

**Reply by StefanOssendorf** [reply]
Sorry I missed your response.
The only NRE in `VerifyAuthorizationCache` is `ApplicationContext` being `null`.
Do you retrieve the data via `Fetch`? Are you using a remote data portal (e.g. asp.net core API endpoint)?

**Reply by fundez** [reply]
What Grid are you using? I had to abandon Telerik and move to Devexpress. Telerik was trying to add a copy of the object being edited without a data portal and didn't provide an event to intercept the add.

**Reply by ecaldentey** [reply]
@fundez,

I'm using DevExpress Blazor 24.1.3
